pandas/core/internals/managers.py

```diff
@@ -341,9 +380,22 @@ def setitem(self: T, indexer, value) -> T:
         if isinstance(indexer, np.ndarray) and indexer.ndim > self.ndim:
             raise ValueError(f"Cannot set values with ndim > {self.ndim}")
 
+        if not self._has_no_reference(0):
```

Comment:
This might be block manager knowledge I'm not aware of, but why can this always be the `0` column?

Reply:
So this method is only ever called when there is a single block (but this can still be because it is either a Series, or a DataFrame with a single block, I think, and thus should not be moved to SingleBlockManager). 
I will update the comment.